### PR TITLE
feat(core): add GLWE linear algebra primitives

### PR DESCRIPTION
**Pull request by IceTDrinker**
closes: https://github.com/zama-ai/tfhe-rs-internal/issues/33

### PR content/description

- add appropriate tests and doctest

### Check-list:

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [x] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [x] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

File: tfhe/src/core_crypto/algorithms/test/glwe_linear_algebra.rs [hunk 1]
```diff
@@ -0,0 +1,776 @@
...
+    let delta: Scalar = encoding_with_padding / msg_modulus;
+
+    while msg != Scalar::ZERO {
+        msg = msg.wrapping_sub(Scalar::ONE);
+        for _ in 0..NB_TESTS {
+            let glwe_sk = allocate_and_generate_new_binary_glwe_secret_key(
+                glwe_dimension,
+                polynomial_size,
+                &mut rsc.secret_random_generator,
+            );
```
Comment (mayeul-zama) [inline tfhe/src/core_crypto/algorithms/test/glwe_linear_algebra.rs:25]: This could be in a common generic functions which take a closure as input

Comment (IceTDrinker) [reply tfhe/src/core_crypto/algorithms/test/glwe_linear_algebra.rs:25]: not really, parameters required per tests vary and we don't always query all of them, so it kind of defeats the purpose of the generic function ?

Comment (IceTDrinker) [reply tfhe/src/core_crypto/algorithms/test/glwe_linear_algebra.rs:25]: I'm not sure

Comment (mayeul-zama) [reply tfhe/src/core_crypto/algorithms/test/glwe_linear_algebra.rs:25]: Not sure either, gotta test to know :man_shrugging:

Comment (IceTDrinker) [reply tfhe/src/core_crypto/algorithms/test/glwe_linear_algebra.rs:25]: I'm a bit wary about genericizing test code, it was painful in concrete-core, and sometimes when the test code is not fully laid out it's a pain to debug, so unless you think it's absolutely necessary I'm not too kin on changing that and am fine with repeating myself in the tests

File: tfhe/src/core_crypto/entities/cleartext.rs [hunk 2]
```diff
@@ -1,7 +1,79 @@
...
+    }
+}
+
+impl<T: Numeric> From<CleartextRefMut<'_, T>> for Cleartext<T>
+where
+    T: Copy,
+{
+    fn from(cleartext_ref: CleartextRefMut<T>) -> Cleartext<T> {
+        Cleartext(*cleartext_ref.0)
+    }
```
Comment (tmontaigu) [inline tfhe/src/core_crypto/entities/cleartext.rs:78]: Technically, I think implementing ToOwned would be a better fit
https://doc.rust-lang.org/std/borrow/trait.ToOwned.html#required-methods

Comment (IceTDrinker) [reply tfhe/src/core_crypto/entities/cleartext.rs:78]: good point, it's just the pattern we had for Plaintext (that I also did 🙄 )

File: tfhe/src/core_crypto/algorithms/glwe_linear_algebra.rs [hunk 3]
```diff
@@ -0,0 +1,1045 @@
...
+{
+    let mut body = lhs.get_mut_body();
+    let ciphertext_modulus = body.ciphertext_modulus();
+    assert!(ciphertext_modulus.is_compatible_with_native_modulus());
+
+    if ciphertext_modulus.is_native_modulus() {
+        body.as_mut()
+            .iter_mut()
+            .zip(rhs.iter())
+            .for_each(|(dst_body, src_plain)| *dst_body = (*dst_body).wrapping_add(*src_plain.0));
```
Comment (tmontaigu) [inline tfhe/src/core_crypto/algorithms/glwe_linear_algebra.rs:290]: Would `slice_wrapping_add_assign(body.as_mut(), rhs.as_ref())` be possible ?

File: tfhe/src/core_crypto/algorithms/glwe_linear_algebra.rs [hunk 4]
```diff
@@ -0,0 +1,1045 @@
...
+{
+    let mut body = lhs.get_mut_body();
+    let ciphertext_modulus = body.ciphertext_modulus();
+    assert!(ciphertext_modulus.is_compatible_with_native_modulus());
+
+    if ciphertext_modulus.is_native_modulus() {
+        body.as_mut()
+            .iter_mut()
+            .zip(rhs.iter())
+            .for_each(|(dst_body, src_plain)| *dst_body = (*dst_body).wrapping_sub(*src_plain.0));
```
Comment (tmontaigu) [inline tfhe/src/core_crypto/algorithms/glwe_linear_algebra.rs:390]: Same idea, `slice_wrapping_sub_assign(body.as_mut(), rhs.as_ref()) be possible ?

File: tfhe/src/core_crypto/algorithms/glwe_linear_algebra.rs (hunk 3 above)
Comment (IceTDrinker) [reply tfhe/src/core_crypto/algorithms/glwe_linear_algebra.rs:290]: most likely yes, not sure what I was thinking there 🙄

File: tfhe/src/core_crypto/entities/cleartext.rs (hunk 2 above)
Comment (IceTDrinker) [reply tfhe/src/core_crypto/entities/cleartext.rs:78]: ah damn forgot to update this -_-

File: tfhe/src/core_crypto/algorithms/glwe_linear_algebra.rs [hunk 5]
```diff
@@ -0,0 +1,1045 @@
...
+    InCont: ContainerMut<Element = Scalar>,
+{
+    let mut body = lhs.get_mut_body();
+    let ciphertext_modulus = body.ciphertext_modulus();
+    assert!(ciphertext_modulus.is_compatible_with_native_modulus());
+
+    if ciphertext_modulus.is_native_modulus() {
+        body.as_mut()
+            .iter_mut()
+            .for_each(|dst_body| *dst_body = (*dst_body).wrapping_add(rhs.0));
```
Comment (tmontaigu) [inline tfhe/src/core_crypto/algorithms/glwe_linear_algebra.rs:488]: slice_wrapping_scalar_add_assign

File: tfhe/src/core_crypto/algorithms/glwe_linear_algebra.rs [hunk 6]
```diff
@@ -0,0 +1,1045 @@
...
+    InCont: ContainerMut<Element = Scalar>,
+{
+    let mut body = lhs.get_mut_body();
+    let ciphertext_modulus = body.ciphertext_modulus();
+    assert!(ciphertext_modulus.is_compatible_with_native_modulus());
+
+    if ciphertext_modulus.is_native_modulus() {
+        body.as_mut()
+            .iter_mut()
+            .for_each(|dst_body| *dst_body = (*dst_body).wrapping_sub(rhs.0));
```
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/core_crypto/algorithms/glwe_linear_algebra.rs:583]: that one does note exist yet 😄 I'll add it then

File: tfhe/src/core_crypto/algorithms/glwe_linear_algebra.rs (hunk 4 above)
Comment (IceTDrinker) [reply tfhe/src/core_crypto/algorithms/glwe_linear_algebra.rs:390]: definitely

File: tfhe/src/core_crypto/entities/cleartext.rs (hunk 2 above)
Comment (IceTDrinker) [reply tfhe/src/core_crypto/entities/cleartext.rs:78]: is it fine to have both ?

Comment (tmontaigu) [reply tfhe/src/core_crypto/entities/cleartext.rs:78]: A yes Borrow requires to return a ref so can't really use it

**Review by tmontaigu — APPROVED** [review]

**Pull request merged**